osu.Server.DifficultyCalculator/Program.cs

```diff
@@ -0,0 +1,214 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using McMaster.Extensions.CommandLineUtils;
+using MySql.Data.MySqlClient;
+using osu.Game.Beatmaps.Legacy;
+using osu.Game.Rulesets.Mania.Mods;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Server.DifficultyCalculator
+{
+    [Command]
+    public class Program
+    {
+        public static void Main(string[] args)
+            => CommandLineApplication.Execute<Program>(args);
+
+        // Todo: Move to configuration
+        private const string connection_string = "Server=localhost;Database=osu;User ID=root;SslMode=None;Pooling=true; Min pool size = 50; Max pool size = 200; Charset=utf8;";
+        private const string beatmaps_path = "osu";
+
+        private readonly Dictionary<string, int> attributeIds = new Dictionary<string, int>();
+
+        private Database database;
+
+
+        public void OnExecute(CommandLineApplication app, IConsole console)
+        {
+            database = new Database(connection_string);
+
+            using (var reader = database.RunQuery("SELECT attrib_id, name FROM osu_difficulty_attribs"))
+            {
+                while (reader.Read())
+                    attributeIds[reader.GetString(1)] = reader.GetInt32(0);
+            }
+
+            var tasks = new List<Task>();
+
+            using (var reader = database.RunQuery("SELECT osu_beatmaps.beatmap_id, checksum FROM osu_beatmaps ORDER BY beatmap_id ASC"))
+            {
+                while (reader.Read())
+                {
+                    int id = reader.GetInt32(0);
+                    string hash = reader.IsDBNull(1) ? string.Empty : reader.GetString(1);
+                    tasks.Add(processBeatmap(id));
+                }
+            }
+
+            Task.WaitAll(tasks.ToArray());
+        }
+
+        private async Task processBeatmap(int beatmapId)
+        {
+            string path = Path.Combine(beatmaps_path, beatmapId + ".osu");
+            if (!File.Exists(path))
+                return;
+
+            await Task.Run(async () =>
+            {
+                var localBeatmap = new LocalWorkingBeatmap(path);
+
+                // Todo: For each ruleset
+                var playable = localBeatmap.GetPlayableBeatmap(localBeatmap.BeatmapInfo.Ruleset);
+                var ruleset = localBeatmap.BeatmapInfo.Ruleset.CreateInstance();
+
+                foreach (var mod in ruleset.GetModsFor(ModType.DifficultyCalculation))
+                {
+                    await Task.Run(() =>
+                    {
+                        var legacyMod = toLegacyMod(mod);
+
+                        var attributes = new Dictionary<string, object>();
+                        double starRating = ruleset.CreateDifficultyCalculator(playable, toModArray(mod)).Calculate(attributes);
+
+                        database.RunNonQuery(
```

Comment:
Database connections are not thread-safe like this. you will need to create a new connection per task. Connection pooling is handled automatically (based on connection string) so there's no harm in `using (new MysqlConnection())` each time.

Reply:
A new connection is still created with every query (see usages of `getConnection()` in `Database.cs`.